circuits/circuits/claiming/distribute.circom

```diff
@@ -0,0 +1,229 @@
+            root <== sorted_root
+        );
+
+        // Prove that the ith leaf in the sorted tree exists in the filtered tree
+        for (var j = 0; j < claim_levels; j++) {
+            hack_2[i][j] <-- filtered_pathElements[reverse_permutation_Indices[i]][j];
+        }
+        CheckMerkleProofStrict(claim_levels)(
+            leaf <== sorted_leaves[i],
+            pathElements <-- hack_2[i], // pathElements can be unconstrained because a merkle proof with a given [leaf, index, root] proof only succeeds for one set of path elements
+            pathIndices <== Num2Bits(claim_levels)(reverse_permutation_Indices[i]),
+            root <== filtered_root
+        );
+    }
+
+    // ---- Step 3 ----
+    // Verify that sorted tree is sorted, and contains no repitition except 0 leaves at the end
+    // "hack" signal is workaround for https://github.com/iden3/circom/issues/189
+    signal sorted_claims_hack[2 ** claim_levels][3];
+    for (var i = 0; i < 2 ** claim_levels; i++) {
+        for (var j = 0; j < 3; j++) {
+            sorted_claims_hack[i][j] <-- claims[reverse_permutation_Indices[i]][j];
+        }
+    }
+
+    for (var i = 1; i < 2 ** claim_levels; i++) {
+        ClaimsInOrder()(
+            leaves <== [sorted_leaves[i-1], sorted_leaves[i]],
+            claims <-- [sorted_claims_hack[i-1], sorted_claims_hack[i]] // claims is advice that is constrained to correspond to the leaf in the component
+            // TODO: pass claims in like this
+            // claims <-- [claims[reverse_permutation_Indices[i-1]], claims[reverse_permutation_Indices[i]]]
+        );
+    }
+}
+
+// ClaimsInOrder verifies that a list of leaves strictly ascends as follows, where H is a hash function:
+// H([0-10]), H([12-15]) ... H([1001-1005]), 0, 0, ..., 0
+//
+// This requires two checks on each pair of elements:
+//      1/ Zeroes must be followed by zeroes
+//      2/ Non-zero values must be ordered
+// 
+// CAUTION: when we move to Nova, we will probably halt before checking every possible leaf, so we must make sure that the unchecked leaves are also zero.
+// This can be done by checking that for some zero leaf, all pathElements where the pathIndex is 0 (i.e., our element is on the left) have the appropriate value for
+// their level. We can cheaply precalculate the proper value as p_k = h(p_{k+1}, p_{k+1}) where p_n = 0 for a depth n tree by caching each p_k.
+template ClaimsInOrder() {
+    signal input leaves[2];
+    // claims is unconstrained advice, since we conditionally check that the claim hashes to the leaf in the component
+    // It needs to be passed the correct preimages for non-zero nodes or witness generation will fail
+    signal input claims[2][3]; 
+
+    // --- Step 1 ----
+    // If we have a zeroed value, make sure it's followed by a zero
+    // i.e., if the previous value is zero, the current value can't be non-zero: NAND(prev_is_zero, curr_is_nonzero)
+    signal curr_is_zero <== IsZero()(leaves[1]);
```

Comment:
rename to `is_curr_zero` and same for the `curr_is_nonzero` and `is_prev_zero`.

Reply:
Awww man, I feel like that's stretching the convention too far. Like, `IsEqual` makes grammatical sense in a sentence like "`a` IsEqaul to `b`," but I just don't see how `is_curr_nonzero` really fits into a nice sentence.

Reply:
ok, i take `is` as a way to denote that it's a boolean and not something that has to be grammatically correct. but it's ok if you don't see it that way. resolving.